Add tree-sitter and tree-sitter-langs

ubolonton opened the pull request:
### Brief summary of what the package does

Incremental parsing for Emacs.

- `tree-sitter`: Core APIs, minor modes for syntax highlighting, tree/query debugging.
- `tree-sitter-langs`: Bundle of grammars and syntax highlighting queries for some common languages.

### Direct link to the package repository

https://github.com/ubolonton/emacs-tree-sitter

### Your association with the package

I'm the maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [ ] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### Notes
- `checkdoc` is not entirely happy, but its suggestions seem pedantic/inaccurate.
- These packages download pre-built binaries upon installation, from GitHub (dynamic module) and bintray (grammars).

riscy commented:
Thanks for these -- I'm doing a quick first pass.  The most important thing to note is that simply loading a package shouldn't cause Emacs to do a bunch of heavy lifting (such as downloading files, etc.)  The heavy lifting should happen after the user explicitly begins interacting with the package.

I'll focus on just the tree-sitter recipe for now (since the other has it as a requirement):

### tree-sitter--mac-load.el ###

byte-compile (using Emacs 27.1):
- tree-sitter--mac-load.el#L34: Don't set `no-byte-compile: t`.
  Byte-compiling is enabled in what follows
- Consider using lexical binding.
  See: https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org#fixing-typical-problems
- Lexical binding will be used in what follows
- No issues!

checkdoc (using version 0.6.2):
```
Warning (emacs): 
tree-sitter--mac-load.el:0: The first line should be of the form: ";;; package --- Summary"
Warning (emacs): 
tree-sitter--mac-load.el:0: You should have a section marked ";;; Commentary:"
Warning (emacs): 
tree-sitter--mac-load.el:10: You should have a section marked ";;; Code:"
Warning (emacs): 
tree-sitter--mac-load.el:35: The footer should be: (provide 'tree-sitter--mac-load)\n;;; tree-sitter--mac-load.el ends here
Warning (emacs): 
tree-sitter--mac-load.el:11: All variables and subroutines might as well have a documentation string
```

package-lint (using version 20200901.2204):
```
2 issues found:
1:0: error: There is no `provide' form.
10:0: error: "ts--try-load-dyn" doesn't start with package's prefix "tree-sitter".
```

Other possible lints:
- tree-sitter--mac-load.el#L31: `fmakunbound` should only _very rarely_ occur in packages

### tree-sitter-cli.el ###

byte-compile (using Emacs 27.1):
- No issues!

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
- No issues!


### tree-sitter-core.el ###

byte-compile (using Emacs 27.1):
```
tree-sitter-core.el:18:1:Error: github.com/443 Temporary failure in name resolution
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
```
25 issues found:
31:0: error: "ts--without-restriction" doesn't start with package's prefix "tree-sitter".
38:0: error: "ts--save-context" doesn't start with package's prefix "tree-sitter".
48:0: error: "ts-point-from-position" doesn't start with package's prefix "tree-sitter".
56:0: error: "ts--point-from-position" doesn't start with package's prefix "tree-sitter".
68:0: error: "ts-point-to-position" doesn't start with package's prefix "tree-sitter".
83:0: error: "ts--buffer-input" doesn't start with package's prefix "tree-sitter".
98:0: error: "ts--buffer-substring-no-properties" doesn't start with package's prefix "tree-sitter".
106:0: error: "ts--node-text" doesn't start with package's prefix "tree-sitter".
115:0: error: "ts-node-text" doesn't start with package's prefix "tree-sitter".
123:0: error: "ts-get-descendant-for-position-range" doesn't start with package's prefix "tree-sitter".
130:0: error: "ts-get-named-descendant-for-position-range" doesn't start with package's prefix "tree-sitter".
137:0: error: "ts-get-child-by-field" doesn't start with package's prefix "tree-sitter".
143:0: error: "ts-node-start-position" doesn't start with package's prefix "tree-sitter".
147:0: error: "ts-node-end-position" doesn't start with package's prefix "tree-sitter".
151:0: error: "ts-node-position-range" doesn't start with package's prefix "tree-sitter".
158:0: error: "ts-goto-first-child-for-position" doesn't start with package's prefix "tree-sitter".
163:0: error: "ts-lang-field-id" doesn't start with package's prefix "tree-sitter".
169:0: error: "ts-lang-node-type-id" doesn't start with package's prefix "tree-sitter".
181:0: error: "ts--stringify-patterns" doesn't start with package's prefix "tree-sitter".
192:0: error: "ts-make-query" doesn't start with package's prefix "tree-sitter".
208:0: error: "ts-query-matches" doesn't start with package's prefix "tree-sitter".
226:0: error: "ts-query-captures" doesn't start with package's prefix "tree-sitter".
247:0: error: "ts-pp-to-string" doesn't start with package's prefix "tree-sitter".
251:0: error: "ts--node-steps" doesn't start with package's prefix "tree-sitter".
276:0: error: "ts--node-from-steps" doesn't start with package's prefix "tree-sitter".
```

Other possible lints:
- tree-sitter-core.el#L14: It's safer to sharp-quote function names; use `#'`

### tree-sitter-debug.el ###

byte-compile (using Emacs 27.1):
```
tree-sitter-debug.el:15:1:Error: github.com/443 Temporary failure in name resolution
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
- No issues!


### tree-sitter-dyn-get.el ###

byte-compile (using Emacs 27.1):
- No issues!

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
- No issues!


### tree-sitter-extras.el ###

byte-compile (using Emacs 27.1):
```
tree-sitter-extras.el:14:1:Error: github.com/443 Temporary failure in name resolution
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
```
4 issues found:
42:18: error: You should depend on (emacs "26.1") if you need `pixel-posn-y-at-point'.
44:11: error: You should depend on (emacs "26.1") if you need `pixel-scroll-pixel-down'.
45:9: error: You should depend on (emacs "26.1") if you need `pixel-scroll-pixel-up'.
67:31: error: You should depend on (emacs "26.1") if you need `pixel-posn-y-at-point'.
```


### tree-sitter-hl.el ###

byte-compile (using Emacs 27.1):
```
tree-sitter-hl.el:14:1:Error: github.com/443 Temporary failure in name resolution
```

checkdoc (using version 0.6.2):
```
Warning (emacs): 
tree-sitter-hl.el:575: Disambiguate font-lock-mode by preceding w/ function,command,variable,option or symbol.
```

package-lint (using version 20200901.2204):
- No issues!


### tree-sitter-load.el ###

byte-compile (using Emacs 27.1):
```
tree-sitter-load.el:16:1:Error: github.com/443 Temporary failure in name resolution
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
- No issues!


### tree-sitter-query.el ###

byte-compile (using Emacs 27.1):
```
tree-sitter-query.el:15:1:Error: github.com/443 Temporary failure in name resolution
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
- No issues!

Other possible lints:
- tree-sitter-query.el#L91: It's safer to sharp-quote function names; use `#'`
- tree-sitter-query.el#L102: It's safer to sharp-quote function names; use `#'`
- tree-sitter-query.el#L103: It's safer to sharp-quote function names; use `#'`

### tree-sitter-tests.el ###

byte-compile (using Emacs 27.1):
- tree-sitter-tests.el#L455: Don't set `no-byte-compile: t`.
  Byte-compiling is enabled in what follows
```
tree-sitter-tests.el:13:1:Error: github.com/443 Temporary failure in name resolution
```
You will probably want to explicitly `:exclude` this file from your recipe.

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
```
63 issues found:
29:0: error: "ts-test-make-parser" doesn't start with package's prefix "tree-sitter".
36:0: error: "ts-test-full-path" doesn't start with package's prefix "tree-sitter".
42:0: error: "ts-test-tree-sexp" doesn't start with package's prefix "tree-sitter".
51:0: error: "ts-test-use-lang" doesn't start with package's prefix "tree-sitter".
61:0: error: "ts-test-with" doesn't start with package's prefix "tree-sitter".
67:0: error: "ts-test-with-file" doesn't start with package's prefix "tree-sitter".
75:0: error: "ts-test-lang-with-file" doesn't start with package's prefix "tree-sitter".
83:0: error: "creating-parser" doesn't start with package's prefix "tree-sitter".
86:0: error: "language::info" doesn't start with package's prefix "tree-sitter".
86:0: error: `language::info' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
91:0: error: "language::equality" doesn't start with package's prefix "tree-sitter".
91:0: error: `language::equality' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
98:0: error: "language::node-types" doesn't start with package's prefix "tree-sitter".
98:0: error: `language::node-types' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
114:0: error: "language::fields" doesn't start with package's prefix "tree-sitter".
114:0: error: `language::fields' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
129:0: error: "parsing::rust-string" doesn't start with package's prefix "tree-sitter".
129:0: error: `parsing::rust-string' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
139:0: error: "parsing::without-setting-language" doesn't start with package's prefix "tree-sitter".
139:0: error: `parsing::without-setting-language' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
144:0: error: "parsing::rust-buffer" doesn't start with package's prefix "tree-sitter".
144:0: error: `parsing::rust-buffer' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
160:0: error: "minor-mode::basic-editing" doesn't start with package's prefix "tree-sitter".
160:0: error: `minor-mode::basic-editing' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
175:0: error: "minor-mode::incremental:change-case-region" doesn't start with package's prefix "tree-sitter".
175:0: error: `minor-mode::incremental:change-case-region' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
185:0: error: "minor-mode::incremental:delete-non-ascii-text" doesn't start with package's prefix "tree-sitter".
185:0: error: `minor-mode::incremental:delete-non-ascii-text' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
193:0: error: "minor-mode::node-at-point" doesn't start with package's prefix "tree-sitter".
193:0: error: `minor-mode::node-at-point' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
207:0: error: "node::eq" doesn't start with package's prefix "tree-sitter".
207:0: error: `node::eq' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
215:0: error: "node::using-without-tree" doesn't start with package's prefix "tree-sitter".
215:0: error: `node::using-without-tree' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
225:0: error: "node::types" doesn't start with package's prefix "tree-sitter".
225:0: error: `node::types' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
249:0: error: "cursor::walk" doesn't start with package's prefix "tree-sitter".
249:0: error: `cursor::walk' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
257:0: error: "cursor::reset" doesn't start with package's prefix "tree-sitter".
257:0: error: `cursor::reset' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
266:0: error: "cursor::using-without-tree" doesn't start with package's prefix "tree-sitter".
266:0: error: `cursor::using-without-tree' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
273:0: error: "conversion::position<->ts-point" doesn't start with package's prefix "tree-sitter".
273:0: error: `conversion::position<->ts-point' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
286:0: error: "buffer-input::non-ascii-characters" doesn't start with package's prefix "tree-sitter".
286:0: error: `buffer-input::non-ascii-characters' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
293:0: error: "buffer-input::narrowing" doesn't start with package's prefix "tree-sitter".
293:0: error: `buffer-input::narrowing' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
300:0: error: "query::making" doesn't start with package's prefix "tree-sitter".
300:0: error: `query::making' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
313:0: error: "query::basic" doesn't start with package's prefix "tree-sitter".
313:0: error: `query::basic' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
335:0: error: "query::range-restriction" doesn't start with package's prefix "tree-sitter".
335:0: error: `query::range-restriction' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
362:0: error: "load" doesn't start with package's prefix "tree-sitter".
366:0: error: "hl::extend-region" doesn't start with package's prefix "tree-sitter".
366:0: error: `hl::extend-region' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
379:0: error: "hl::hl-region-vs-query-region" doesn't start with package's prefix "tree-sitter".
379:0: error: `hl::hl-region-vs-query-region' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
397:0: error: "hl::face-mapping" doesn't start with package's prefix "tree-sitter".
397:0: error: `hl::face-mapping' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
435:0: error: "hl::bench" doesn't start with package's prefix "tree-sitter".
435:0: error: `hl::bench' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
```


### tree-sitter.el ###

byte-compile (using Emacs 27.1):
```
tree-sitter.el:20:1:Error: github.com/443 Temporary failure in name resolution
```
As mentioned, simply loading (or even just byte compiling a file) shouldn't do things like make network requests.

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200901.2204):
```
1 issue found:
199:0: error: "turn-on-tree-sitter-mode" doesn't start with package's prefix "tree-sitter".
```
I haven't looked, but this looks like you want to use Emacs facilities for defining major/minor modes.

Other possible lints:
- tree-sitter.el#L260: Quoting this lambda may prevent it from being compiled
- tree-sitter.el#L260: It's safer to sharp-quote function names; use `#'`

### Loadability ###

Verifying ability to #'load each file:
```
Loading tree-sitter--mac-load.el
Loading tree-sitter-cli.el
Loading tree-sitter-core.el
tree-sitter-core.el:Error: Emacs 27.1 errored during load
Loading tree-sitter-debug.el
tree-sitter-debug.el:Error: Emacs 27.1 errored during load
Loading tree-sitter-dyn-get.el
Loading tree-sitter-extras.el
tree-sitter-extras.el:Error: Emacs 27.1 errored during load
Loading tree-sitter-hl.el
tree-sitter-hl.el:Error: Emacs 27.1 errored during load
Loading tree-sitter-load.el
tree-sitter-load.el:Error: Emacs 27.1 errored during load
Loading tree-sitter-query.el
tree-sitter-query.el:Error: Emacs 27.1 errored during load
Loading tree-sitter-tests.el
tree-sitter-tests.el:Error: Emacs 27.1 errored during load
Loading tree-sitter.el
tree-sitter.el:Error: Emacs 27.1 errored during load
Done.
```
Note that these are all failing to load because Emacs wasn't given network access:
```
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
Eager macro-expansion failure: (error "github.com/443 Temporary failure in name resolution")
Downloading https://github.com/ubolonton/emacs-tree-sitter/releases/download/0.10.0/tree-sitter-dyn.so
```

### Package ###

- Prefer the default recipe if possible.
- Add a LICENSE file that GitHub can detect (e.g. no markup) if possible
  See: https://github.com/licensee/licensee
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter--mac-load.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-cli.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-core.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-debug.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-dyn-get.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-extras.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-hl.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-load.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-query.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter-tests.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) to tree-sitter.el
- Cargo.lock -- not elisp
- Cargo.toml -- not elisp
- lisp/tree-sitter--mac-load.el (unknown license) -- (no header)
- lisp/tree-sitter-cli.el (unknown license) -- Utilities for tree-sitter CLI
- lisp/tree-sitter-core.el (unknown license) -- Core tree-sitter APIs
- lisp/tree-sitter-debug.el (unknown license) -- Debug tools for tree-sitter
- lisp/tree-sitter-dyn-get.el (unknown license) -- Utilities to obtain tree-sitter-dyn
- lisp/tree-sitter-extras.el (unknown license) -- Extra functionalities of tree-sitter
- lisp/tree-sitter-hl.el (unknown license) -- Syntax highlighting based on tree-sitter
- lisp/tree-sitter-load.el (unknown license) -- Language loading for tree-sitter
- lisp/tree-sitter-query.el (unknown license) -- Tools for running queries live
- lisp/tree-sitter-tests.el (unknown license) -- Tests for tree-sitter.el
- lisp/tree-sitter.el (unknown license) -- Incremental parsing system
- src -- directory

ubolonton commented:
> The most important thing to note is that simply loading a package shouldn't cause Emacs to do a bunch of heavy lifting (such as downloading files, etc.) The heavy lifting should happen after the user explicitly begins interacting with the package.

The files to be downloaded are essential parts of the package. They are platform-specific binaries, which MELPA currently doesn't have a satisfactory mechanism for distribution. We can discuss such a mechanism when it becomes a big-enough pain point (when there are enough packages that use dynamic modules). Until then, this is the most reasonable workaround.

> tree-sitter--mac-load.el#L31: fmakunbound should only very rarely occur in packages

`tree-sitter--mac-load.el` is a load-time workaround for a dynamic module's tech debt on Emacs. The function it defines should be unbound once the it's done. I may try refactoring it to use `require` and `unload-feature`.

> tree-sitter-core.el#L14: It's safer to sharp-quote function names; use #'

AFAICT, not sharp-quoting the argument of `functionp` is the norm.

> 42:18: error: You should depend on (emacs "26.1") if you need `pixel-posn-y-at-point'.

These are optional, and already conditionally handled.

> tree-sitter-query.el#L91: It's safer to sharp-quote function names; use #'

Yeah, these should be sharp-quoted. I'll fix them.

> 31:0: error: "ts--without-restriction" doesn't start with package's prefix "tree-sitter".

The `ts-` prefix is intentional, to distinguish between core functionalities and app-level features, as explained [here](https://ubolonton.github.io/emacs-tree-sitter/api/). It's not named `ts.el` due to the [conflict with the timestamp handling lib](https://github.com/ubolonton/emacs-tree-sitter/issues/35). Until we can decide on a better prefix, it will remain `tree-sitter-core.el`.

> You will probably want to explicitly :exclude this file from your recipe.

Good idea! I'll do that.

> I haven't looked, but this looks like you want to use Emacs facilities for defining major/minor modes.

Yeah. `turn-on-foo` seems to be the conventional name for the helper function that conditionally turns on a globalized mode.

> tree-sitter.el#L260: It's safer to sharp-quote function names; use #'

Not sharp-quoting here is intentional, to avoid creating an env-capturing closure. What are the alternatives?

riscy commented:
Everything you wrote is reasonable.  For this one:

> The ts- prefix is intentional, to distinguish between core functionalities and app-level features, as explained here. It's not named ts.el due to the conflict with the timestamp handling lib. Until we can decide on a better prefix, it will remain tree-sitter-core.el.

Unfortunately these names are the only way we have to namespace packages and prevent difficult debugging situations from reaching users.  You may not have any collisions _now_ (or ever) but it is far easier at scale for MELPA to standardize on this rule.  Can you change `ts-` to `tree-sitter-core`?  Alternatively, if you really need a short prefix, you could add a dependency package to MELPA called, e.g., `tsc`, that only contains the tree-sitter-core functions.

> ... What are the alternatives?

Sorry, that was a false positive. :) 

I'll reiterate these since my first review was noisy:

- Add a LICENSE file that GitHub can detect (e.g. no markup) if possible
  See: https://github.com/licensee/licensee
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter--mac-load.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-cli.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-core.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-debug.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-dyn-get.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-extras.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-hl.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-load.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter-query.el
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to tree-sitter.el

ubolonton commented:
> Alternatively, if you really need a short prefix, you could add a dependency package to MELPA called, e.g., `tsc`, that only contains the tree-sitter-core functions.

Is it ok to include a `tsc.el` file (or maybe `ets.el`/`lts.el`) in the `tree-sitter` package itself? Splitting it into another package would be quite involved in terms of changes to the CI pipelines. AFAICT, for ease of debugging, it's more important for the functions to start with the containing file's prefix, rather than the package's prefix as well.

riscy commented:
> Is it ok to include a tsc.el file (or maybe ets.el/lts.el) in the tree-sitter package itself, for now? ... AFAICT, for ease of debugging, it's more important for the functions to start with the containing file's prefix, rather than the package's prefix as well.

The filenames can be whatever you want (except that there must be at least one "package-main" file whose name matches your packages's name), but the functions inside them should still be prefixed with your package's name.  package-lint can check for this right now, although it requires [some configuration](https://github.com/purcell/package-lint/blob/master/package-lint.el#L47-L55), and I'm also checking it on my end.  For MELPA, it's more important for the functions to start with the package's prefix.

> Splitting it into another package would be quite involved in terms of changes to the CI pipelines.

Noted, and splitting is sort of our worst case scenario here and probably not a good idea unless "ets" (or whatever) is an independently useful package.

ubolonton commented:
I'll close this for now. I'll reopen it after I have time to split the core part into another package.

ubolonton commented:
I've split the core part into another package, `tsc`. It contains all the core APIs, prefixed with `tsc-`.

I also added `SPDX-License-Identifier` to the Lisp files, and removed `tree-sitter--mac-load.el`.

riscy commented:
Thanks a lot for accommodating.  One more nag about the license, and then this is an easy merge.

- Add a LICENSE file that GitHub can detect (e.g. no markup) if possible
  See: https://github.com/licensee/licensee
- This is not a blocker, but I would recommend sharp-quoting the function names in tsc-obsolete.el to get the compiler checking those for you

ubolonton commented:
I've added the LICENSE file.

`tsc.el` depends on `tsc-obsolete.el`, so the functions haven't been defined when `tsc-obsolete.el` is loaded. Not sharp-quoting avoids the unnecessary compiler warnings.

riscy commented:
Thank you!
